Add vnetif tests for connect and iperf + backend for trailing bytes

MagnusS opened the pull request:
Adds two new tests:

- test_connect connects two network stacks and sends a short hello world string
- test_iperf runs iperf between two network stacks

Both tests are run with the basic mirage-vnetif backend, as well as a new backend that adds random trailing bytes.

Depends on [mirage-vnetif](https://github.com/MagnusS/mirage-vnetif)

avsm commented:
superb!

> On 30 Apr 2015, at 15:15, Magnus Skjegstad <notifications@github.com> wrote:
> 
> Adds two new tests:
> 
> test_connect connects two network stacks and sends a short hello world string
> test_iperf runs iperf between two network stacks
> Both tests are run with the basic mirage-vnetif backend, as well as a new backend that adds random trailing bytes.
> 
> Depends on mirage-vnetif <https://github.com/MagnusS/mirage-vnetif>
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/mirage/mirage-tcpip/pull/129 <https://github.com/mirage/mirage-tcpip/pull/129>
> Commit Summary
> 
> Add vnetif tests for connect and iperf, w/wo trailing bytes
> File Changes
> 
> M _oasis <https://github.com/mirage/mirage-tcpip/pull/129/files#diff-0> (2)
> M _tags <https://github.com/mirage/mirage-tcpip/pull/129/files#diff-1> (22)
> A lib_test/common.ml <https://github.com/mirage/mirage-tcpip/pull/129/files#diff-2> (18)
> M lib_test/test.ml <https://github.com/mirage/mirage-tcpip/pull/129/files#diff-3> (4)
> A lib_test/test_connect.ml <https://github.com/mirage/mirage-tcpip/pull/129/files#diff-4> (72)
> A lib_test/test_iperf.ml <https://github.com/mirage/mirage-tcpip/pull/129/files#diff-5> (156)
> A lib_test/vnetif_backends.ml <https://github.com/mirage/mirage-tcpip/pull/129/files#diff-6> (38)
> A lib_test/vnetif_common.ml <https://github.com/mirage/mirage-tcpip/pull/129/files#diff-7> (54)
> M myocamlbuild.ml <https://github.com/mirage/mirage-tcpip/pull/129/files#diff-8> (4)
> M setup.ml <https://github.com/mirage/mirage-tcpip/pull/129/files#diff-9> (13)
> Patch Links:
> 
> https://github.com/mirage/mirage-tcpip/pull/129.patch <https://github.com/mirage/mirage-tcpip/pull/129.patch>
> https://github.com/mirage/mirage-tcpip/pull/129.diff <https://github.com/mirage/mirage-tcpip/pull/129.diff>
> —
> Reply to this email directly or view it on GitHub <https://github.com/mirage/mirage-tcpip/pull/129>.
>

MagnusS commented:
Mirage-vnetif is submitted to opam, but not merged yet ([PR](https://github.com/ocaml/opam-repository/pull/4034)) - so Travis will fail for now

samoht commented:
Can you `ocp-indent` the files?

yomimono commented:
I can only run the tests if tcpip is already installed and has made `dlltcpip_stubs` available somewhere -- otherwise, `test.byte` fails to load a shared library:

```
$ make test
ocaml setup.ml -build 
W: Cannot find source file matching module 'tcp' in library tcp
Finished, 0 targets (0 cached) in 00:00:00.
Finished, 248 targets (226 cached) in 00:00:02.
ocaml setup.ml -test 
Fatal error: cannot load shared library dlltcpip_stubs
Reason: dlltcpip_stubs.so: cannot open shared object file: No such file or directory
W: Test 'test' fails: Command '/home/user/mirage-tcpip/_build/lib_test/test.byte' terminated with error code 2
E: Failure("Tests had a 100.00% failure rate")
make: *** [test] Error 1
```

Presumably we'd want to run against the version of this library that's just been built, rather than whatever's in ~/.opam?

yomimono commented:
(also I'm very excited about the presence of these tests)

samoht commented:
> `Fatal error: cannot load shared library dlltcpip_stubs`

This is because the tests needs to be compiled with `-custom`, see https://github.com/mirage/ocaml-cstruct/blob/master/_oasis#L75

avsm commented on lib_test/vnetif_common.ml:
```diff
@@ -0,0 +1,54 @@
+    or_error "tcpv4" S.T.connect ipv4 >>= fun tcpv4 ->
+    let config = {
+        V1_LWT.name = "stack";
+        V1_LWT.console = c; 
```
Just one `V1_LWT` needed here; rest are redundant in the record as it is inferred.

MagnusS commented:
Thanks! I've changed the following:
- Added -custom to _oasis and updated
- Cleaned up the code in common.ml based on Thomas' suggestions 
- Removed Lwt.syntax dependency
- Removed redundant V1_LWT
- ocp-indent'ed everything

samoht commented:
Yay!
